rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,934 @@
+$ tctl autoupdate run --group staging-group
+Executing auto-update for group 'staging-group' immediately.
+```
+
+```yaml
+kind: autoupdate_version
+spec:
+  # agent_version is the version of the agent the cluster will advertise.
+  agent_version: X.Y.Z
+  # agent_version_schedule specifies the rollout schedule associated with the version.
+  # Currently, only critical, regular, and immediate schedules are permitted.
+  agent_version_schedule: regular|critical|immediate
+
+  # ...
+```
+
+```shell
+$ tctl autoupdate update --set-agent-version=15.1.1
+Automatic updates configuration has been updated.
+$ tctl autoupdate update --set-agent-version=15.1.2 --critical
+Automatic updates configuration has been updated.
+```
+
+Notes:
+- `autoupdate_version` is separate from `autoupdate_config` so that Cloud customers can be restricted from updating `autoupdate_version`, while maintaining control over the rollout.
+
+#### Rollout
+
+```yaml
+kind: agent_rollout_plan
+spec:
+  # start time of the rollout
+  start_time: 0001-01-01T00:00:00Z
+  # target version of the rollout
+  version: X.Y.Z
+  # schedule that triggered the rollout
+  schedule: regular
+  # hosts updated by the rollout
+  host_count: 127
+status:
+  # current host index in rollout progress
+  last_active_host_index: 23
+  # failed hosts
+  failed_host_count: 3
+  # timed-out hosts
+  timeout_host_count: 1
+```
+
+Notes:
+- This resource is stored in a paginated format with separate keys for each page and progress
+
+### Version Promotion
+
+This RFD only proposed a mechanism to signal when agent auto-updates should occur.
+Advertising different target Teleport versions for different groups of agents is out-of-scope for this RFD.
+This means that groups which employ auto-scaling or ephemeral resources will slowly converge to the latest Teleport version.
+
+**This could lead to a production outage, as the latest Teleport version may not receive any validation before it is advertised to newly provisioned resources in production.**
+
+To solve this in the future, we can use the group name (provided to `/v1/webapi/find` and specified via `teleport-update enable`) to determine which version should be served.
+
+This will require tracking the desired version of groups in the backend, which will add additional complexity to the rollout logic.
+
+## Details - Linux Agents
+
+We will ship a new auto-updater package for Linux servers written in Go that does not interface with the system package manager.
+It will be distributed as a separate package from Teleport, and manage the installation of the correct Teleport agent version manually.
+It will read the unauthenticated `/v1/webapi/find` endpoint from the Teleport proxy, parse new fields on that endpoint, and install the specified agent version according to the specified update plan.
+It will download the correct version of Teleport as a tarball, unpack it in `/var/lib/teleport`, and ensure it is symlinked from `/usr/local/bin`.
```

Comment:
The binaries are stored in separate version-managed directories, such that only symlinks are overwritten.

We leave behind only the last working version, so that we can rollback quickly. See:
https://github.com/gravitational/teleport/pull/40190/files#diff-4fb277bd3d884c7b045e7efad61111ac4e7957c063e0c0c9281f4e74fd72997fR524